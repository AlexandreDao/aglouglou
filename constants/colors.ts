export const BACKGROUND_COLOR = '#431327'

export const TAB_BAR_BACKGROUND_COLOR = '#380F1F'

export const ACTIVE_COLOR = '#FF8EB6'

export const INACTIVE_COLOR = '#A7989D'

export const TEXT_COLOR = 'white'

export const SEPARATOR_COLOR = 'white'
