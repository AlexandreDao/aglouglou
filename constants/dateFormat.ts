export enum DateFormat { 
  DATE_TIME = 'YYYY-MM-DD HH:mm:ss',
}