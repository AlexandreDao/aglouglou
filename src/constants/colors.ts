export const BACKGROUND_COLOR = '#431327'

export const TAB_BAR_BACKGROUND_COLOR = '#380F1F'

export const ACTIVE_COLOR = '#FF8EB6'

export const INACTIVE_COLOR = '#A7989D'

export const TEXT_COLOR = '#EDEEF0'

export const SEPARATOR_COLOR = 'white'

export const ALCOHOLIC_CATEGORY_COLOR = '#13432F'

export const DRINK_CATEGORY_COLOR = '#acab00'

export const BACKDROP_COLOR = 'rgba(0, 0, 0, 0.4)'

export const ANDROID_RIPPLE_COLOR = 'rgba(255, 255, 255, 0.6)'
