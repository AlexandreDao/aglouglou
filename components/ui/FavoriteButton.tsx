import { Pressable, Alert, StyleSheet } from 'react-native'
import React from 'react'
import { IconSymbol } from './IconSymbol'
import { ACTIVE_COLOR, INACTIVE_COLOR } from '@/constants/colors'
import { CocktailDetail } from '@/types/Cocktail'
import { useAppDispatch } from '@/hooks/store/useAppDisptach'
import { addToFavorite, removeFromFavorite } from '@/store/slices/favoritesSlice'
import Animated, { useAnimatedStyle, useSharedValue, withSpring, withTiming } from 'react-native-reanimated'

interface FavoriteButtonProps {
  isFavorite: boolean
  favorite: () => void
  unfavorite: () => void
}

const styles = StyleSheet.create({
  pressable: {
    padding: 16,
  }
})

function FavoriteButton({isFavorite, favorite, unfavorite}: FavoriteButtonProps) {
  const rotation = useSharedValue(0)
  const yPosition = useSharedValue(0)
  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{ rotateY: `${rotation.value}deg` }, { translateY: yPosition.value }],
    };
  });

  return (
    <Pressable 
      style={styles.pressable} 
      onPress={() => {
         if (isFavorite) {
            unfavorite()
          }
          else {
            rotation.value = withTiming(360, { duration: 600 }, () => {
              rotation.value = 0;
            })
            yPosition.value = withSpring(-20, { duration: 600 }, () => {
              yPosition.value = 0;
            })
            favorite()
          }
      }}
    >
      <Animated.View style={animatedStyle}>
        <IconSymbol name="star.fill" size={32} color={isFavorite ? ACTIVE_COLOR : INACTIVE_COLOR} />
      </Animated.View>
    </Pressable>
  )
}

export default FavoriteButton