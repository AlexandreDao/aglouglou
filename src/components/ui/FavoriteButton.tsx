import { StyleSheet } from 'react-native'
import React from 'react'
import { IconSymbol } from '@/components/ui/IconSymbol'
import { ACTIVE_COLOR, INACTIVE_COLOR } from '@/constants/colors'
import Animated, { useAnimatedStyle, useSharedValue, withSpring, withTiming } from 'react-native-reanimated'
import { Pressable } from 'react-native-gesture-handler'

interface FavoriteButtonProps {
  isFavorite: boolean
  favorite?: () => void
  unfavorite?: () => void
  size?: number
}

const styles = StyleSheet.create({
  pressable: {
    alignItems: 'center',
    height: 48,
    justifyContent: 'center',
    width: 48,
  },
})

const FavoriteButton = ({ isFavorite, favorite, unfavorite, size = 32 }: FavoriteButtonProps) => {
  const rotation = useSharedValue(0)
  const yPosition = useSharedValue(0)
  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{ rotateY: `${rotation.value}deg` }, { translateY: yPosition.value }],
    }
  })

  return (
    <Pressable
      testID="favorite-button"
      style={styles.pressable}
      // Prevent event propagation
      onStartShouldSetResponder={() => true}
      onPress={() => {
        if (isFavorite) {
          unfavorite?.()
        } else {
          if (favorite) {
            rotation.value = withTiming(360, { duration: 600 }, () => {
              rotation.value = 0
            })
            yPosition.value = withSpring(-20, { duration: 600 }, () => {
              yPosition.value = 0
            })
            favorite()
          }
        }
      }}
    >
      <Animated.View style={animatedStyle}>
        <IconSymbol name="star.fill" size={size} color={isFavorite ? ACTIVE_COLOR : INACTIVE_COLOR} />
      </Animated.View>
    </Pressable>
  )
}

export default FavoriteButton
