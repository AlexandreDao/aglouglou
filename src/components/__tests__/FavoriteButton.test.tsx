import FavoriteButton from '@/components/FavoriteButton'
import { NOOP } from '@/constants/function'
import { fireEvent, render } from '@testing-library/react-native'

describe('<FavoriteButton />', () => {
  test('FavoriteButton snapshot', () => {
    const tree = render(<FavoriteButton isFavorite={false} unfavorite={NOOP} favorite={NOOP} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('FavoriteButton press favorite', () => {
    const favoriteMock = jest.fn()
    const unfavoriteMock = jest.fn()
    const { getByTestId } = render(
      <FavoriteButton isFavorite={false} unfavorite={unfavoriteMock} favorite={favoriteMock} />
    )

    fireEvent.press(getByTestId('favorite-button'))
    expect(favoriteMock).toHaveBeenCalled()
    expect(unfavoriteMock).not.toHaveBeenCalled()
  })

  test('FavoriteButton press unfavorite', () => {
    const favoriteMock = jest.fn()
    const unfavoriteMock = jest.fn()
    const { getByTestId } = render(<FavoriteButton isFavorite unfavorite={unfavoriteMock} favorite={favoriteMock} />)

    fireEvent.press(getByTestId('favorite-button'))
    expect(unfavoriteMock).toHaveBeenCalled()
    expect(favoriteMock).not.toHaveBeenCalled()
  })
})
