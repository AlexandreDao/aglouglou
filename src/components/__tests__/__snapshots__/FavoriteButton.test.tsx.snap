// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavoriteButton /> FavoriteButton snapshot 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  enabled={true}
  hitSlop={
    {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
  rippleColor={0}
  style={
    [
      {},
      {
        "alignItems": "center",
        "height": 48,
        "justifyContent": "center",
        "width": 48,
      },
    ]
  }
  testID="favorite-button"
>
  <View
    style={
      {
        "transform": [
          {
            "rotateY": "0deg",
          },
          {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <ViewManagerAdapter_SymbolModule
      animated={false}
      colors={[]}
      name="star.fill"
      resizeMode="scaleAspectFit"
      style={
        [
          {
            "height": 24,
            "width": 24,
          },
          [
            {
              "height": 32,
              "width": 32,
            },
            undefined,
          ],
        ]
      }
      tintColor="#A7989D"
      type="monochrome"
      weight="regular"
    />
  </View>
</RNGestureHandlerButton>
`;
